pkg/lightning/backend/external/engine.go

```diff
@@ -362,9 +362,10 @@ func (e *Engine) loadRangeBatchData(ctx context.Context, jobKeys [][]byte, outCh
 	// we shouldn't handle duplicates for OnDuplicateKeyIgnore, it's the semantic
 	// of OnDuplicateKeyError, but to make keep compatible with the old code, we
 	// keep this behavior.
+	// For OnDuplicateKeyError, which is used by adding unique index, we don't return the 'duplicate key error' here.
+	// Instead, we return the error in memoryDataDupDetectIter when writing to tikv
 	// TODO: remove this when we have have fully integrated the OnDuplicateKey.
-	if k := dupKey.Load(); k != nil &&
-		(e.onDup == engineapi.OnDuplicateKeyIgnore || e.onDup == engineapi.OnDuplicateKeyError) {
+	if k := dupKey.Load(); k != nil && e.onDup == engineapi.OnDuplicateKeyIgnore {
```

Comment:
where OnDuplicateKeyError is handled?

Reply:
[`48086f2` (#60848)](https://github.com/pingcap/tidb/pull/60848/commits/48086f220c796a84772eb1469969d03c1a4c1d12)